TASK: Level up to phpstan 2 (Flow 9 adjustments)

mhsdesign opened the pull request:
<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

The upgrade to phpstan level two was introduced via https://github.com/neos/flow-development-collection/pull/3264, this holds the flow 9 specific adjustments.

Related (phpstan level 1) https://github.com/neos/flow-development-collection/pull/3216

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->


**Review instructions**


<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked wit `!!!` and have upgrade-instructions

kitsunet commented on Neos.Flow/Classes/Cli/Response.php:
```diff
@@ -47,12 +47,12 @@ class Response
-     * @var
+     * @var bool
      */
     private $colorSupport;
```
I guess we could also start actually typehinting those? they are private anyways.

mhsdesign replied:
yes we could ... but i have resisted any temptation to use real types to prevent runtime errors or shoot people in the foot via inheritance. @mficzel and me also thought that it would also be the easiest to review ;)

eventually we should migrate of course ^^

kitsunet commented on Neos.Flow/Classes/Core/Booting/Scripts.php:
```diff
@@ -577,8 +578,10 @@ public static function initializeSystemFileMonitor(Bootstrap $bootstrap)
             }
-            foreach ($package->getAutoloadPaths() as $autoloadPath) {
-                self::monitorDirectoryIfItExists($fileMonitors['Flow_ClassFiles'], $autoloadPath, '\.php$');
+            if ($package instanceof GenericPackage) {
```
nothing you need to fix here, but we should probably add getAutoloadPaths() to the PackageInterface...

kitsunet commented on Neos.Flow/Classes/Property/PropertyMappingConfiguration.php:
```diff
@@ -361,11 +366,10 @@ public function traverseProperties(array $splittedPropertyPath)
                 $this->subConfigurationForProperty[$currentProperty] = clone $this->subConfigurationForProperty[self::PROPERTY_PATH_PLACEHOLDER];
             } else {
-                $this->subConfigurationForProperty[$currentProperty] = new $type;
+                $this->subConfigurationForProperty[$currentProperty] = new static();
```
nice one!

mficzel commented on Neos.Eel/Classes/FlowQuery/OperationInterface.php:
```diff
@@ -42,7 +42,7 @@ public static function isFinal();
      * to implement runtime conditions.
      *
-     * @param array (or array-like object) $context onto which this operation should be applied
+     * @param $context array (or array-like object) $context onto which this operation should be applied
```
Should'nt that be `@param array $context`?

mficzel replied:
or `@param array|iterable $context`

mhsdesign replied:
yes good catch ;)

mficzel commented on Neos.Eel/Classes/FlowQuery/Operations/AbstractOperation.php:
```diff
@@ -80,7 +80,7 @@ public static function getShortName()
      * {@inheritdoc}
      *
-     * @param array (or array-like object) $context onto which this operation should be applied
+     * @param $context array (or array-like object) $context onto which this operation should be applied
```
`@param array|iterable $context`

mficzel commented on Neos.Flow/Classes/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php:
```diff
@@ -249,14 +249,7 @@ public function loadMetadataForClass($className, ClassMetadata $metadata)
-                    } else {
-                        $primaryTable['uniqueConstraints'][] = $uniqueConstraint;
-                    }
+                    $primaryTable['uniqueConstraints'][] = $uniqueConstraint;
```
Isnt that a change in bahavior as now `name` is not used as array index anymore? Also this would also transfer other keys like `fields` from the constraint. Might well be valid but i would like to unterstand the motivation.

mhsdesign replied:
The problem was dead code. Phpstan (rightfully) detected that we foreach over `$uniqueConstraint` but in 251 we reassign the variable and thus the check `if (!empty($uniqueConstraint->options)) {` will always fail

the code was introduced as fix via https://github.com/neos/flow-development-collection/commit/012f0e76a3169ed9b30b2dd23698718c4d550782 but it never worked. So removing its seems the right choice? - I have no clue btw about this file ^^

mhsdesign replied:
Extracted this less simple part into an own pr https://github.com/neos/flow-development-collection/pull/3263

mficzel commented on Neos.FluidAdaptor/Classes/Core/Parser/Interceptor/ResourceInterceptor.php:
```diff
@@ -91,7 +91,9 @@ public function setDefaultPackageKey($defaultPackageKey)
-        /** @var $node TextNode */
+        if (!$node instanceof TextNode) {
+            return $node;
+        }
```
Why is that non breaky? Would unterstand the if-instanceof around the next lines but not before

mhsdesign replied:
The next line always assumes its a TextNode `$node->getText()` but this might fail, as this is a generic processor. Instead when we encounter a node which is not a TextNode we will just early return instead of causing an error.

mficzel commented on Neos.FluidAdaptor/Classes/View/TemplatePaths.php:
```diff
@@ -356,20 +358,6 @@ protected function sanitizePath($path)
-
-        return Files::concatenatePaths([$packageResourcesPath, 'Private']);
-    }
-
```
Why is this removed specifically? Looks unused but so do other methods in there.

mhsdesign replied:
yes its unused and phpstan doesnt like it. Other unused methods might be public which is okay.

mficzel reviewed (commented):
Generally like this very much, left some questions

mhsdesign reviewed (commented):
.

mhsdesign committed:
TASK: Remove obsolete internal method `TemplatePaths::getPackagePrivateResourcesPath`

mhsdesign committed:
TASK: `Neos\FluidAdaptor\Core\Parser\Interceptor\ResourceInterceptor::process` early return

Previously it was assumed that `$node` is always a `TextNode`, as the following code implies, by calling for example `getText`.
This might fail, as this is a generic processor.
And for full correctness we will just early return instead of causing an error.

mhsdesign commented:
This is now ready to be merged. The tests are now completely green with phpstan level 2.

**NOTE FOR REVIEWERS**
Changes which do not seem self explanatory have been extracted into separate commits with in depth information.
Adding uncritical php type doc comments has been squashed into one commit.

mhsdesign commented:
Thänks @kitsunet ❤️

mhsdesign commented:
Fyi i also created a backport for Flow 8.3 https://github.com/neos/flow-development-collection/pull/3264

bwaidelich reviewed (approved):
+1 by reading

mhsdesign commented:
i think it would make sense to merge the backport first to not mess to much with the history: https://github.com/neos/flow-development-collection/pull/3264

The backport basically contains all changes of `TASK: **Pling pling** level up to phpstan 2` as cherry-pick plus an adjustment commit `TASK: Backport phpstan level 2 to Flow 8.3 and ignore to be fixed things` to make things work on 8.3

mhsdesign committed:
TASK: Find a working workaround for phpstan to accept this weird situation

This must be cleaned up at some point:
See https://github.com/neos/flow-development-collection/issues/3272

mhsdesign committed:
TASK: **Pling pling** level up to phpstan 2 - Flow 9 adjustments

mhsdesign committed:
TASK: Revert type ignore hack for XMLReader and phpstan

As the issue is fixed: https://github.com/phpstan/phpstan/issues/8629
